update steps for container tutorial (#850)
LLNL/benchpark: docs/example-workflow.rst

   SPDX-License-Identifier: Apache-2.0

===================================================
Hello Benchpark Tutorial: Kripke Benchmark Example
===================================================

--------------------------------------
Step 1: Verify Benchpark Installation
--------------------------------------

First we will ensure Benchpark is installed and working correctly::

>     benchpark --version
> 
> This should return a version number like ``0.1.0``.

Reviewer: ```suggestion
    benchpark --version

.. program-output:: ../bin/benchpark --version
```
Should be this to get version dynamically

    ==>   Executing 4 out of 4 experiments:
    ==>   Log files for experiments are stored in: /home/jovyan/benchpark/wkp/kripke-benchmark/tutorial/workspace/logs/execute.2025-07-09_18.14.08
    ==> Running executors...
    ƒV54uD5o5  
    ƒV57fKkEK  
    ƒV5ANUS6s  
    ƒV5D498h5

.. note::

    You can check on the progress of your runs by running the command::

        flux jobs [-a]

    The -a is optional; adding it to the command will show all the jobs submitted including jobs that are currently running and completed jobs.

The experiments will run sequentially, and the total time to complete all four experiments should be 8-9 minutes. Upon completion, each
experiment will generate a caliper file. When all the experiments are finished, running the command :code:`flux jobs -a` will show all the
jobs. A completed job will be green with ``CD`` as its status (Example below):

    .. image:: ./finished_experiments_example.png
       :alt: Example output of flux jobs -a when experiments are finished
       :width: 750px
       :align: center

------------------------
Step 7: Analyze Results
------------------------

1. **Conduct** pre-defined analysis with Benchpark::

>     benchpark analyze --workspace-dir /home/jovyan/benchpark/wkp/kripke-benchmark/tutorial/workspace

Reviewer: @stephanielam3211  Try:
1. `--no-mpi`
2. `--no-mpi` and `--top-n-regions 10`
3. You can try `--chart-fontsize` to change chart fontsize

   SPDX-License-Identifier: Apache-2.0

==================================================
Hello Benchpark Tutorial: Kripke Benchmark Example
==================================================

**Objective:**

This tutorial will guide you through using Benchpark to run a strong scaling
experiment with the `Kripke benchmark <https://github.com/LLNL/Kripke>`_ on an AWS instance. By the end of
this tutorial, you will be able to use Benchpark to:

* Initialize the details of a system and scaling experiment with Benchpark
* Run the scaling experiment with Ramble

> * Perform pre-defined performance analysis on the results of the scaling experiment with Benchpark

Reviewer: ```suggestion
* Initialize a system configuration and experiment configuration
* Build and run a scaling experiment
* Perform pre-defined analysis on the results of the scaling experiment
```

**Prerequisites:**

> * Access to a terminal with Benchpark installed (provided automatically by the infrastructure in our `Benchpark Tutorial repository <https://github.com/llnl/benchpark-tutorial>`_)

Reviewer: Should this be more specific like `- You need Docker`, etc?

* Basic familiarity with command-line interfaces

--------------------------------------
Step 1: Verify Benchpark Installation
--------------------------------------

First, ensure Benchpark is installed and working correctly by running:

.. code-block:: bash

    benchpark --version

You should see a version number like ``0.1.0``.

----------------------------------------------------
Step 2: Explore Available Benchmarks and Experiments
----------------------------------------------------

Next, list all available benchmarks and experiments by running:

.. code-block:: bash

    benchpark list experiments

You should see an output like:

.. code-block:: text

    Experiments:
        ...
        hpcg+single_node
        hpcg+openmp
        hpcg+strong
        hpcg+weak
        hpl+single_node
        hpl+openmp
        hpl+strong
        hpl+weak
        ior+single_node
        ior+strong
        ior+weak
        kripke+single_node
        kripke+openmp
        kripke+cuda
        kripke+rocm
        laghos+single_node
        laghos+cuda
        laghos+rocm
        lammps+single_node
        lammps+openmp
        lammps+cuda
        lammps+rocm
        lammps+strong

>         ...

Reviewer: I think in general the format of

1. command
2. "you should see"
3. output

could be refactored into

1. $ command
2. output

for all of these commands, e.g.:

```suggestion
    $ benchpark list experiments
    Experiments:
        ...
        hpcg+single_node
        hpcg+openmp
        hpcg+strong
        hpcg+weak
        hpl+single_node
        hpl+openmp
        hpl+strong
        hpl+weak
        ior+single_node
        ior+strong
        ior+weak
        kripke+single_node
        kripke+openmp
        kripke+cuda
        kripke+rocm
        laghos+single_node
        laghos+cuda
        laghos+rocm
        lammps+single_node
        lammps+openmp
        lammps+cuda
        lammps+rocm
        lammps+strong
        ...
```

Reviewer 2: I get what you're saying here, but making this change would break with the structure of the rest of the page. I would not change this.

Author: I like the suggestion, but I think it is ok as is for this venue.

From this output, you can see that Benchpark experiments are represented
with Spack-like specifications (i.e., specs). For example, the spec
``kripke+single_node`` describes an experiment using the Kripke benchmark
running on a single node.

Additionally, you can get only the experiments associated with a particular
benchmark by adding :code:`--experiment <experiment_name>` to the above command.
For example, to get only the experiments associated with Kripke, run:

.. code-block:: bash

    benchpark list experiments --experiment kripke

You should see the following:

.. code-block:: text

    Experiments:
        kripke+single_node
        kripke+openmp
        kripke+cuda
        kripke+rocm

.. note::

    For Kripke, the default experiment is ``kripke+single_node``. Since single node runs are enabled by default, the ``+single_node``
    variant will be excluded from the commands below. Instead, the instructions below that use experiment specs will

>     simply use ``kripke`` instead of ``kripke+single_node``.

Reviewer: `+single_node` is mutually exclusive with `+strong`, so the below experiment is actually `+strong` not single node

Reviewer: For Kripke, we specify the strong scaling experiment on the command line using ``kripke scaling=strong``.

Reviewer 2: Then why does Benchpark print `kripke+single_node` as an experiment if the actual experiment is `kripke scaling=strong`? Something's not adding up between what you're saying and what Benchpark produces, @michaelmckinsey1. I can try rewriting if I understand what the mismatch is.

.. _step3_label:
------------------------------------------
Step 3: Initialize Your System Description
------------------------------------------

Next, initialize the description of the AWS system by running the commands below:

.. code-block:: bash

   cd benchpark
   benchpark system init --dest=hpdc-tutorial aws-tutorial instance_type=c7i.12xlarge

> The :code:`benchpark system init` command generates Ramble and Spack configuration

Reviewer: ```suggestion
The :code:`benchpark system init` command generates configuration
```

files that describe the system on which you are running. The system is specified using

> a Spack-like specification (i.e., spec). In the command above, the spec (i.e., :code:`aws-tutorial instance_type=c7i.12xlarge`)

Reviewer: We use `Spack-like` multiple times. Is this an intentional point? I think `a specification` would be sufficient.

Reviewer 2: This is absolutely intentional. Lots of people know Spack, but very few people know Benchpark. By saying these specs are "Spack-like", we're relating them to a concept they are already familiar with, which can help them understand faster.

* ``software.yaml``: a Ramble configuration file specifying the default packages to use for software like compilers and MPI
* ``variables.yaml``: a Ramble configuration file defining variables that are needed for job script generation and scheduling (e.g., type of scheduler, number of cores per node)
* ``auxiliary_software_files/compilers.yaml``: a Spack configuration file defining available compilers on the system
* ``auxiliary_software_files/packages.yaml``: a Spack configuration file defining available software on the system

----------------------------------
Step 4: Initialize Your Experiment
----------------------------------

Next, initialize the Kripke strong scaling experiment used in this tutorial by running:

.. code-block:: bash

    benchpark experiment init --dest=kripke-benchmark kripke scaling=strong caliper=time,mpi

Similar to :code:`benchpark system init`, the :code:`benchpark experiment init` command generates
the Ramble configuration file to describe the experiment to be run. The experiment is specified
using a Spack-like specification (i.e., spec). In the command above, the spec (i.e., :code:`kripke scaling=strong caliper=time,mpi`)

> defines a strong-scaling experiment running Kripke with `Caliper <https://github.com/llnl/caliper`>_ support. The spec also enables Caliper's time measurement with MPI

Reviewer: Clarification
```suggestion
defines a strong-scaling experiment running Kripke with the `Caliper <https://github.com/llnl/caliper`>_ profiling tool enabled. The ``caliper=time,mpi`` specification enables (1) measuring time and (2) automated profiling of MPI regions.
```

> support at runtime.

Reviewer: This is relevant to the above change.
```suggestion
```

> After running the command above, you should see a ``ramble.yaml`` file in the ``kripke-benchmark`` directory.

Reviewer: Clarification
```suggestion
After running the command above, you should see a Ramble configuration file (``ramble.yaml``) in the ``kripke-benchmark`` directory.
```

and :code:`benchpark system init` (i.e., ``hpdc-tutorial/``) and combines them to generate a Benchpark workspace.
A Benchpark workspace contains everything that Benchpark, Ramble, and Spack need to build and run your experiment, including:

* Clones of Spack and Ramble
* A ``setup.sh`` script that calls Spack and Ramble's setup scripts
* A Ramble workspace

To start using your Benchpark workspace, run:

.. code-block:: bash

    . /home/jovyan/benchpark/wkp/setup.sh

.. _step6_label:
-----------------------------------------------------------------
Step 6: Build Software Dependencies and Generate Experiment Files
-----------------------------------------------------------------

Next, build any necessary software and generate all necessary files for the Kripke scaling experiment by running:

.. code-block:: bash

    ramble --disable-progress-bar \
    --workspace-dir /home/jovyan/benchpark/wkp/kripke-benchmark/hpdc-tutorial/workspace \
    workspace setup

This command does two things. First, it builds all necessary software using Spack. Building the software may take a while to complete, depending

> on how many external packages are contained in the system definition from :ref:`Step 3 <step3_label>`. For this tutorial, it should take roughly 5 minutes.

Reviewer: Not sure if it's necessary to mention externals here, or maybe rephrase. Almost sounds like more externals makes it take longer.
```suggestion
This command does two things. First, it builds all necessary software using Spack. Building the software may take a while to complete. For this tutorial, it should take roughly 5 minutes.
```

Second, this command configures files (e.g. submission scripts) needed to perform the runs that make up the current experiment.

> For each run in the experiment, a directory containing the files necessary for the run will be created under ``/home/jovyan/benchpark/wkp/kripke-benchmark/hpdc-tutorial/workspace/experiments/kripke/kripke``. If the command is successful, you should see something like:

Reviewer: Rewording
```suggestion
Second, this command generates batch scripts for executing the experiment.
For each run in the experiment, a directory containing the files necessary for the run will be created under ``/home/jovyan/benchpark/wkp/kripke-benchmark/hpdc-tutorial/workspace/experiments/kripke/kripke/``. If the command is successful, you should see something like:
```

    ==> Experiment #4 (4/4):
    ==>     name: kripke.kripke.kripke_kripke_single_node_strong_scaling_caliper_time_mpi_4_4_2_64_64_32_64_1_128_128_4_32
    ==>     root experiment_index: 4
    ==>     log file: /home/jovyan/benchpark/wkp/kripke-benchmark/hpdc-tutorial/workspace/logs/setup.2025-07-09_18.08.23/kripke.kripke.kripke_kripke_single_node_strong_scaling_caliper_time_mpi_4_4_2_64_64_32_64_1_128_128_4_32.out
    ==>   Returning to log file: /home/jovyan/benchpark/wkp/kripke-benchmark/hpdc-tutorial/workspace/logs/setup.2025-07-09_18.08.23.out

------------------------------------------
Step 7: Run Kripke Experiment using Ramble
------------------------------------------

Next, run the Kripke strong scaling experiment by running the following command:

.. code-block:: bash

    ramble --disable-progress-bar \
    --workspace-dir /home/jovyan/benchpark/wkp/kripke-benchmark/hpdc-tutorial/workspace \
    on

This command simply passes the submission scripts generated in :ref:`Step 6 <step6_label>` to the system's resource manager (which is specified
in the files generated by :code:`benchpark system init`). For the AWS

> infrastructure used in this tutorial, the resource manager is LLNL's `Flux resource manager <https://flux-framework.org/>`_.

Reviewer: Rewording
```suggestion
This command submits the batch scripts generated in :ref:`Step 6 <step6_label>` to the system's resource manager (which is specified
in the files generated by :code:`benchpark system init`). For the AWS
infrastructure used in this tutorial, the resource manager is LLNL's `Flux resource manager <https://flux-framework.org/>`_.
```

    only one job will run at a time under our infrastructure because each user only has 1 node. If you are running on an HPC system,
    expect the jobs to complete faster.

After all the jobs are finished, each job directory (i.e., subdirectories of ``/home/jovyan/benchpark/wkp/kripke-benchmark/hpdc-tutorial/workspace/experiments/kripke/kripke``)
will contain a Caliper output file (i.e., a ``.cali`` file) containing performance data for the job.

------------------------
Step 8: Analyze Results
------------------------

Finally, we perform pre-defined analysis on the Caliper files generated by the
scaling study we defined in the experiment. The ``benchpark analyze`` command uses
`Thicket <https://github.com/llnl/thicket>`_ to compose and visualize the Caliper data
and generate scaling charts. We want to generate a chart of the experiment, focusing on
on the application-level function calls, we specify the ``--no-mpi`` flag to hide MPI
function calls in the resulting graph. We also specify the ``--chart-fontsize`` to
increase the overall font size in the resulting graph, helpful for better readability of

> the graph in presentations.

Reviewer: ```suggestion
scaling study we defined in the experiment. The ``benchpark analyze`` command uses
the `Thicket <https://github.com/llnl/thicket>`_ performance analysis tool to compose the Caliper performance profiles and visualize the data using a stacked area chart.
Focusing on the application-level function calls, we specify the ``--no-mpi`` flag to hide MPI function calls in the resulting graph. We also specify ``--chart-fontsize`` to
increase the font size in the resulting graph.
```

.. code-block:: bash

    benchpark analyze \
    --workspace-dir /home/jovyan/benchpark/wkp/kripke-benchmark/hpdc-tutorial/workspace \
    --no-mpi \
    --chart-fontsize 15

The command above reads in the Caliper files generated by the experiment and generates several files, such as the plot and calling context tree shown below.

> These files can be found in ``/home/jovyan/benchpark/wkp/kripke-benchmark/hpdc-tutorial/workspace/analyze``.

Reviewer: Clarification
```suggestion
The command above reads in the Caliper files generated by the experiment and generates several files, such as the stacked area chart figure and Caliper calltree shown below.
These files can be found in ``/home/jovyan/benchpark/wkp/kripke-benchmark/hpdc-tutorial/workspace/analyze``.
```

Hello Benchpark Example
=========================

.. note::

    Make sure you can use the benchpark command; check by typing in::

        benchpark --version

>     You should receive ``0.1.0`` on the terminal

Author: "This should return" instead of "You should receive"

Reviewer 3: fixed

    Add steps to init, setup, build, run analyze on existing system (container
    ideally)

1. Navigate into the benchpark directory::

    cd benchpark

2. Initialize the variant of the AWS system using the existing system specification in Benchpark::

    benchpark system init --dest=tutorial aws-tutorial instance_type=c7i.24xlarge

> 3. To run the openmp, single node scaling version of the Kripke benchmark, initialize it for experiments::

Author: Is this still openmp and single node? I'm not seeing either of these terms listed in the actual experiment init command.

Reviewer 3: I just took it out

    benchpark setup kripke-benchmark/ tutorial/ wkp/

.. note::

    This command configures a Ramble workspace using the system specification from ``benchpark system init``
    and experiment specification from ``benchpark experiment init``. Ramble can then use this workspace to
    build, run, and configure the Kripke benchmark for the AWS System we are using for this tutorial.

Benchpark will provide next steps to the console but they are also provided here. 

5. Run the setup script for dependency software, Ramble and Spack::

    . /home/jovyan/benchpark/wkp/setup.sh

6. Then setup the Ramble experiment workspace::

    ramble --disable-progress-bar --workspace-dir /home/jovyan/benchpark/wkp/kripke-benchmark/tutorial/workspace workspace setup

This command does 2 main things. First, it builds all necessary software using Spack. This process may take some time (2-3 minutes).
Second, this command configures files (e.g. Flux submission script) needed to perform the runs that make up the current experiment.
For each run, a directory will be created under ``benchpark/wkp/kripke-benchmark/tutorial/workspace/experiments/kripke/kripke``. 

> If the setup is successfull, you will see something like this after the setup command::

Author: typo here successfull

Reviewer 3: fixed

    ==>   Returning to log file: /home/jovyan/benchpark/wkp/kripke-benchmark/tutorial/workspace/logs/setup.2025-07-09_18.08.23.out

7. Next, we run the Kripke experiments, which will launch jobs through the `Flux resource manager <https://flux-framework.org/>`_::

    ramble --disable-progress-bar --workspace-dir /home/jovyan/benchpark/wkp/kripke-benchmark/tutorial/workspace on

You should receive something like this on the terminal after running the command::

    ==> Streaming details to log:
    ==>   /home/jovyan/benchpark/wkp/kripke-benchmark/tutorial/workspace/logs/execute.2025-07-09_18.14.08.out
    ==>   Executing 4 out of 4 experiments:
    ==>   Log files for experiments are stored in: /home/jovyan/benchpark/wkp/kripke-benchmark/tutorial/workspace/logs/execute.2025-07-09_18.14.08
    ==> Running executors...
    ƒV54uD5o5  
    ƒV57fKkEK  
    ƒV5ANUS6s  
    ƒV5D498h5

.. note::

    You can check on the progress of your runs by running the command::

>         flux jobs [-a]

Author: why is -a in square brackets? what does this indicate?

Reviewer 3: -a is an optional argument and it lets the user see all the jobs they had submitted including the ones that are pending, currently running, and completed jobs. I added a sentence over it underneath the command

The experiments will run sequentially, and the total time to complete all four experiments should be 8-9 minutes. Upon completion, each
experiment will generate a caliper file. When all the experiments are finished, running the command :code:`flux jobs -a` will show all the
jobs. A completed job will be green with ``CD`` as its status (Example below):

    .. image:: ./finished_experiments_example.png
       :alt: Example output of flux jobs -a when experiments are finished
       :width: 750px
       :align: center

8. After running the experiments, conduct pre-defined analysis with Benchpark::

>     benchpark analyze --workspace-dir ~/benchpark/wkp/kripke-benchmark/tutorial/workspace

Author: Let's use the `/home/jovyan` instead of `~` for consistency

Reviewer 3: did it

> 9. Navigate to ``~/benchpark/wkp/kripke-benchmark/tutorial/workspace/analyze``. This directory will contain the results from the analysis,

Author: Let's use the `/home/jovyan` instead of `~` for consistency

Reviewer 3: did it
